### PR TITLE
Improve ArraySegment's override of GetHashCode

### PR DESCRIPTION
**Pull request by jamesqo**
This changes `ArraySegment`'s implementation of `GetHashCode` to follow the one in [`ValueTuple`](https://github.com/dotnet/corefx/blob/master/src/System.ValueTuple/src/System/ValueTuple/ValueTuple.cs#L248), which is order-sensitive and hopefully has put more thought into it. As a result, `_offset` and `_count` should no longer be treated interchangeably, e.g. the following won't print `True`:

```cs
var array = new byte[1];
var segment = new ArraySegment<byte>(array, 0, 1);
var segment2 = new ArraySegment<byte>(array, 1, 0);
Console.WriteLine(segment.GetHashCode() == segment2.GetHashCode());
```

Addresses #4550 

cc @JonHanna

### CODE REVIEW

**Comment by ellismg** [comment]
/cc @AlexGhiondea 

We probably want to guard this under `FEATURE_CORECLR`

**Comment by AlexGhiondea** [comment]
Yes, we should add this under `FEATURE_CORECLR`.

Was the hash code copied 'as-is' from another place or is this a brand new way of creating hash codes?

**Comment by ellismg** [comment]
> Was the hash code copied 'as-is' from another place or is this a brand new way of creating hash codes?

It was copied from how ValueTuple does it (see the comment above).

**Comment by jamesqo** [comment]
Thanks for the feedback! I've added an `#ifdef` for `FEATURE_CORECLR` to keep the old implementation in the full framework.

**Comment by AlexGhiondea** [comment]
LGTM

**Comment by JonHanna** [comment]
Balancing cost of calculation vs quality of hash is tricky, especially when changing existing code, but this seems to add very little extra cost for the improvement, so LGTM.

**Comment by jamesqo** [comment]
Test Ubuntu x64 Checked Build and Test 
Test OSX x64 Checked Build and Test
Test FreeBSD x64 Checked Build
Test CentOS7.1 x64 Debug Build and Test

**Comment by jamesqo** [comment]
Test OSX x64 Checked Build and Test
Test Ubuntu x64 Checked Build and Test

**Comment by jamesqo** [comment]
Initially added a [no merge] label since as is pointed out in dotnet/corefx#8193, `ValueTuple`'s hashing function may not be so robust itself-- the way it's currently implemented, it can [easily cause collisions](https://github.com/dotnet/corefx/issues/8193#issuecomment-216333540) if the numbers are high enough.

However, I don't think that applies here. For one, `object.GetHashCode` only returns [26 bits](https://github.com/dotnet/coreclr/blob/775003a4c72f0acc37eab84628fcef541533ba4e/src/vm/syncblk.h#L142) total, so e.g. if you run this as a console app:

```cs
for (int i = 0; i < 100; i++)
{
	int hash = new object().GetHashCode();
	Console.WriteLine(hash & 0x04000000);
}
```

it'll print a continuous stream of zeroes. This means that `array.GetHashCode() * 33` can never overflow, and additionally the array's hash code is very unlikely to be equivalent to the offset/count, so the edge case @VSadov mentioned doesn't really apply here.

**TL;DR** Don't worry about the [no merge] label I added then removed, this is safe to be merged. Posting this to avoid any confusion.

File: src/mscorlib/src/System/ArraySegment.cs [hunk 1]
```diff
@@ -112,9 +112,23 @@ public int Count
...
+#if !FEATURE_CORECLR
+            // Old implementation: Xor everything.
+            // This isn't optimal since it treats _offset and _count interchangeably,
+            // but we keep it around in the full framework for compat.
+            return _array.GetHashCode() ^ _offset ^ _count;
+#else
+            // Taken from ValueTuple.CombineHashCodes
+            int hash = _array.GetHashCode();
+            hash = ((hash << 5) + hash) ^ _offset;
+            hash = ((hash << 5) + hash) ^ _count;
```
Comment (JonHanna) [inline src/mscorlib/src/System/ArraySegment.cs]: `(hash << 5) - hash` would be the equivalent of `hash * 31` which has been found to work relatively well as mult-and-xor (or mult-and-add) hashing goes (alas I can't find the literature I once read on this, it still produced much worse hashes than more complicated algorithms with strong avalanche effects, but they lose when it comes to calculation time). (Why it was found to work well is not clear; that it's a prime, that it's a Mersenne number and that both combine to mean its a Mersenne prime have each been offered as the reason, but it seems to do better than other numbers with the same features).

**Comment by VSadov** [comment]
I'd add some bias to the hash to make it more likely it has lower bits. Any small number would work, but 5381 is traditionally used.
Also, assuming that _array.GetHashCode() is a good well-dispersed hash, it can be just xor-mixed. We have to assume that, otherwise we have a bigger problem.  :-)

So it would become
```C#
            // int hash = 5381;
            // hash = ((hash << 5) + hash) ^ _offset;
            int hash = 177573 ^ _offset;                      // same as above
            hash = ((hash << 5) + hash) ^ _count;
            hash ^= _array.GetHashCode();
```


Generally  h * 33 ^ x[i]   is  not the best hash function around. I guess String uses it for compat reasons. It should be good enough here though.

File: src/mscorlib/src/System/ArraySegment.cs (hunk 1 above)
Comment (jamesqo) [inline src/mscorlib/src/System/ArraySegment.cs]: @JonHanna I'm choosing to stick with `33` for now; I've done a bit of Googling on the subject, and it looks like although the latter is composite, 31 and 33 both result in very few collisions. See here: http://math.stackexchange.com/a/59952 and here: https://gist.github.com/hmic/1676398#file-hash-h-L20-L21 As 33 seems to be more idiomatic when calculating Bernstein hashes, I'm sticking with how it's implemented as-is.

One practical advantage of using `31` for hash calculations instead is that it can be [slightly cheaper to calculate on RISC](http://stackoverflow.com/a/35304979/4077294), but I'm not too familiar with the code generation-aspect of CoreCLR, so I don't know if changing this to `(hash << 5) - hash` would lead to any real perf benefits on those platforms.

**Comment by jamesqo** [comment]
@VSadov Thanks for reviewing my changes! I've incorporated the changes you've mentioned, with one exception:

> Also, assuming that _array.GetHashCode() is a good well-dispersed hash, it can be just xor-mixed.

`object.GetHashCode` only returns the [lower 26 bits](https://github.com/dotnet/coreclr/blob/775003a4c72f0acc37eab84628fcef541533ba4e/src/vm/syncblk.h#L142) of the actual hash code (for what reason I don't know why). As a result, this will always print `True`:

```cs
Console.WriteLine((new object().GetHashCode() & 0xFC000000) == 0);
```

What I've decided to do is shift the array's hash to the left by 6 before xoring it with the hash, so those upper bits don't get wasted (since most of the time `offset` and `count` will be low compared to the array's hash code).

File: src/mscorlib/src/System/ArraySegment.cs [hunk 2]
```diff
@@ -112,9 +112,35 @@ public int Count
...
+            // NOTE: (hash << 5) + hash is equal to hash * 33.
+            
+            int hash = (33 * 5381) ^ _offset;
+            hash = ((hash << 5) + hash) ^ _count;
+            
+            int arrayHash = _array.GetHashCode();
+            // Note quite sure why, but it looks like the result of object.GetHashCode
+            // only returns 26 bits in the return value- the upper 6 are set to zero.
+            // See src/vm/syncblk.h for more.
+            // As a result, we take the array hash and shift it 6 to the left
```
Comment (VSadov) [inline src/mscorlib/src/System/ArraySegment.cs]: I do not think we need to make assumptions about object hash taking 26bits. 
It should be treated as a pseudo-random number not correlated with _offset or _count in any ways and as such it could be just directly mixed with a xor.

Comment (VSadov) [inline src/mscorlib/src/System/ArraySegment.cs]: The reason is mostly that it seems such an implementation detail. It may not stay the same in future versions or be the same on other platforms like .net native.

Also, it is not uncommon pattern to just mask N lower bits when N bit hash is needed. In such cases shifting would make the array identity insignificant in small hashes.

Comment (jamesqo) [inline src/mscorlib/src/System/ArraySegment.cs]: @VSadov It is [the same](https://github.com/dotnet/corert/blob/4376c55dd018d8b7b9fed82449728711231ba266/src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs#L53) on .NET Native though, although yeah I agree that it does seem like somewhat of an implementation detail. If it's alright with you, I think we should change the approach to switch the lower 8 bits before XORing with the array, like this:

```cs
int hash = (33 * 5381) ^ _offset;
hash = ((hash << 5) + hash) ^ _count;
hash = (hash << 24) | (hash >> 8); // since _offset and _count are typically low,
// this will move their bits to the top of the integer
hash ^= _array.GetHashCode();
```

This way we don't waste the upper few bits due to `object.GetHashCode`, and still stay resilient in the future if it's changed to return more than 26 bits.

**Comment by davidfowl** [comment]
Why diverge from the full framework version? Do you think this change will make it there eventually? Would people need to ifdef or do runtime light up to account of this behavioral difference or is it obscure enough that nobody should depend on it?

**Comment by jamesqo** [comment]
@davidfowl I added an `#ifdef` for coreclr since @ellismg and @AlexGhiondea requested it. Perhaps in some future version of the full framework, we can add a compatibility switch that defaults to the new implementation (see what was suggested in [#229](https://github.com/dotnet/coreclr/pull/229#issuecomment-216220353)).

> Would people need to ifdef or do runtime light up to account of this behavioral difference or is it obscure enough that nobody should depend on it?

No one *should* be depending on this, as `GetHashCode` comes with no guarantees. Even [MSDN](https://msdn.microsoft.com/en-us/library/system.object.gethashcode(v=vs.110).aspx) says so:

> A hash code is intended for efficient insertion and lookup in collections that are based on a hash table. A hash code is not a permanent value.

However, in the unlikely case an issue comes up, the person affected could just (in most cases) make a new comparer, since most uses of `GetHashCode` is with `Dictionary`/`Set`:

```cs
public sealed class LegacyArraySegmentComparer<T> : IEqualityComparer<ArraySegment<T>>
{
    public override int GetHashCode(ArraySegment<T> segment)
    {
        if (segment.Array == null) return 0;
        return segment.Array.GetHashCode() ^ segment.Offset ^ segment.Count;
    }
}
```

All in all I think this is a safe change to make.

**Comment by VSadov** [comment]
Re ``` #ifdef ``` - I do not think there is a big compat concern here, because of randomization that object.GetHashCode introduces to the result. It is hard to see a meaningful scenario that would break after the change. 
However I'd defer to  @ellismg and @AlexGhiondea to make a call about that.

**Comment by VSadov** [comment]
@jamesqo - I think rotating the object hash before mixing is unnecessary. It relies on implementation detail, and even then it would not improve hash all that much.
Rotating is not very expensive, but with virtually same goodness of the results it would be better to not do it.

Basically - It is not a worthwhile practice in general to rotate object hashes (or manipulate in any other way) before mixing them as a part of composite hash. I do not see why this case is different. Just xor it in.

**Comment by danmosemsft** [comment]
@jamesqo what work remains here? just to get clean CI?

**Comment by jamesqo** [comment]
@danmosemsft Was waiting for #6767 to get merged so I could update this to use the `HashHelpers` class there.

File: src/mscorlib/src/System/ArraySegment.cs [hunk 3]
```diff
@@ -112,9 +112,19 @@ public int Count
...
-            return null == _array
-                        ? 0
-                        : _array.GetHashCode() ^ _offset ^ _count;
+            if (_array == null)
+            {
+                return 0;
+            }
+            
+            int hash = 5381;
+            hash = System.Numerics.Hashing.HashHelpers.Combine(hash, _offset);
```
Comment (jamesqo) [inline src/mscorlib/src/System/ArraySegment.cs]: Unfortunately, the fully-qualified name has to be used since this clashes with `System.Collections.HashHelpers`.

File: src/mscorlib/src/System/ArraySegment.cs [hunk 4]
```diff
@@ -112,9 +112,19 @@ public int Count
...
+            hash = System.Numerics.Hashing.HashHelpers.Combine(hash, _count);
+            
+            // The array hash is expected to be an evenly-distributed mixture of bits,
+            // so rather than adding the cost of another rotation we just xor it.
+            hash ^= _array.GetHashCode();
+            return hash;
```
Comment (jamesqo) [inline src/mscorlib/src/System/ArraySegment.cs]: Removed the `FEATURE_CORECLR` ifdefs, @jkotas you mentioned there was a change in how PRs are ported to desktop @ https://github.com/dotnet/coreclr/pull/6767#issuecomment-254331763

**Comment by jamesqo** [comment]
@jkotas Could you please take a look at merging this now that `HashHelpers` has been added in #6767? Thanks.
